pg_wait_sampling.c

```diff
@@ -865,27 +878,38 @@ pgws_planner_hook(Query *parse,
 				  int cursorOptions,
 				  ParamListInfo boundParams)
 {
-	if (MyProc)
-	{
-		int i = MyProc - ProcGlobal->allProcs;
-		if (!pgws_proc_queryids[i])
-			pgws_proc_queryids[i] = parse->queryId;
-
-	}
+	PlannedStmt *result;
+	int i = MyProc - ProcGlobal->allProcs;
+	if (nesting_level == 0)
+		pgws_proc_queryids[i] = parse->queryId;
 
 	/* Invoke original hook if needed */
-	if (planner_hook_next)
-		return planner_hook_next(parse,
+	nesting_level++;
+	PG_TRY();
+	{
+		if (planner_hook_next)
+			result = planner_hook_next(parse,
 #if PG_VERSION_NUM >= 130000
-				query_string,
+					query_string,
 #endif
-				cursorOptions, boundParams);
-
-	return standard_planner(parse,
+					cursorOptions, boundParams);
+		else
+			result = standard_planner(parse,
 #if PG_VERSION_NUM >= 130000
-				query_string,
+					query_string,
 #endif
-			cursorOptions, boundParams);
+					cursorOptions, boundParams);
+	}
+	PG_CATCH();
+	{
+		nesting_level--;
+		if (nesting_level == 0)
+			pgws_proc_queryids[i] = UINT64CONST(0);
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
```

Comment:
Hm... we need to do the same (decrement nesting_level and reset queryid) for the non-error case.
See https://github.com/postgres/postgres/blob/REL_12_STABLE/contrib/pg_stat_statements/pg_stat_statements.c#L894